### ISSUE TITLE
Feature request: cheat code support

### ISSUE THREAD

**Issue by sammargh** [mrehkopf/sd2snes#39]
It would be nice if SD2SNES would support PAR/GG codes on games. Krikzz has been also implementing a load cheats from txt file on his everdrives as of recent which is extremely handy for lengthy codes. What would be ideal would be to setup a large flat file with identifiers such as the CRC or the header content and use YAML or something similar to have something much easier to manage.

**Comment by mrehkopf** [comment]
Actually there seem to be two different versions of the CHT format. The one that says "FE FC" has 20 character descriptions. There is another one with different values that only has 13 character descriptions. I never saw that one in the wild but zsnes source code suggests its existence.

**Comment by sammargh** [comment]
Maybe it'd be better to use the DSTwo CHT format as it's more human-readable and won't require hex editors to create:
https://gbatemp.net/threads/guide-make-snes-cht-files-for-the-dstwo.263113/

Would require manually converting GG codes but it would at least allow for a single file for all roms. This also would get around the limitation of figuring out which rom to use as it is up to the user to identify or to add a CRC field in the GameInfo portion if you want to limit the usage.

**Comment by sammargh** [comment]
Ah, whoops. Looks like the emulator CATSFC switched to BSNES-style cheats. Pretty clean format, XML driven using SHA256 for rom identification:
```xml
<cartridge sha256="44428a3d1c796fbd41da7620e321c45f11cd80a0e5f4ab8c48177106cb960d77">
<name>Wolverine - Adamantium Rage (USA)</name>
<cheat>
<description>Invincibility</description>
<code>7E00B8FF</code>
</cheat>
</cartridge>
```

**Comment by mrehkopf** [comment]
Checksum driven auto game matching isn't very feasible on the sd2snes. Finding cheats for a given ROM might be ok as you only need to calc the ROM checksum (but still need to scan through the whole cheats file).
Finding a ROM based on a given checksum is impossible, all ROMs would have to be scanned and calculated. Takes hours at best ;)
I think I'll stick with the "one file per ROM" method. I'm not a big fan of the SNES .CHT format anymore though - the description text is so limited, and it doesn't support multiple codes per cheat.
I like a format like this, it's simple to parse and still somewhat editable by hand:
```
7e00b8ff Invincibility
7e0f34ff, 7e0f35ff, 7e0f36ff Highscore On
ec8ad793 Jump
```
I could also stick to an existing format used by a flash cart but I think neither SNES Powerpak nor Super Everdrive actually save codes to card?

**Comment by sammargh** [comment]
I know the super everdrive supports loading cheats from a txt file however it's GG only and pretty limited in how it works. Basically you make a text file with only the GG code in it and the Everdrive loads it. Not very elegant and your solution would work better plus those codes could easily be converted and adapted to your format.

One thing that would be nice but not necessary is if you could have a sub-directory to your roms for the cheats to be loaded from to keep things organized

**Comment by mrehkopf** [comment]
I see. I might release an intermediate version with a "text-file only" cheat interface and of course I'd like to re-use that format for the GUI controlled cheats later. Another proposal, this would be a disabled cheat:
```
#7e00b8ff Invincibility
```
Actually I would regard any line that doesn't start with a parseable hex number as disabled/invalid.
Also I would like to support both GG and PAR codes for manual input. They can be distinguished by the '-' character in the middle.
I also want to allow GG and PAR codes in the GUI but always handle them as PAR codes internally (and save them to the text file that way too).
Oh, and I forgot to mention - I was actually going to put the cheat files in a separate folder (like /sd2snes/cheats) with <romname.cht> or similar.
As I am now going to use my own format for the cheats, do you think it would be better to choose a different file extension, too? .cht is kind of reserved for the original CHT format. Just .txt maybe? But then it's not very clear that the files are meant to contain cheats.... hm.

**Comment by sammargh** [comment]
Probably would be wise to use a different format for the extension so people don't confuse the type. One last consideration which may or may not be possible is to use YAML for the codes as that would make editing them and reading them quite easy for a human as well as parsing but I'm not sure how powerful the FPGA is or if it is even capable of parsing/editing YAML. An example could be as such:

Filename: Example Rom (USA).sfc
Cheat name: cheats/Example Rom (USA).yml
```yaml
---
- Name: Invincibility
  Code:
   AAAA-AAAA
   BBBB-BBBB
  Enabled:
   True
- Name: Infinite Money
  Code:
   12345678
  Enabled:
   False
```

Then to enable/disable it requires changing the boolean flag on each code. 

Also be sure that whatever filename matching you are using is not prone to the 8.3 bug that SRAM lookup has where if you have multiple files it doesn't assume the first of multiple files by not using the long name.

**Comment by mrehkopf** [comment]
YAML seems like a sane choice, very readable (as opposed to say XML) and easy to parse. I think I'll go for that. Also it has been suggested by two people now ;)
I have proper LFN support in place so there should never be any file name mismatches.

**Comment by magu** [comment]
+1 YAML, makes it three.

**Comment by mrehkopf** [comment]
YAML it is, then.

**Comment by mrehkopf** [comment]
The parser/generator will be minimalist, probably even supporting only even a subset of the eYAML parser proposed here: https://github.com/yaml/YAML2/wiki/eYAML
Advanced features of YAML aren't required for storing the cheat data.

**Comment by sammargh** [comment]
Sounds good! When you have the support finalized I'd like to know so I can begin work on some python scripts to handle creating codes. 

Using BSNES' system I should be able to handle implementing codes from existing packs pretty easily for any game regardless of knowing the file name since they're identified by sha256 and reformat codes accordingly.

**Comment by mrehkopf** [comment]
Generic parser is done, now to bolt on the specifics for cheat files.

**Comment by mrehkopf** [comment]
Cheats via YAML files are implemented since v0.1.7. Stay tuned for the cheat menu in v0.1.8 ;)
